core/src/layers/prometheus.rs

```diff
@@ -93,47 +89,123 @@ use crate::*;
 ///     println!("## Prometheus Metrics");
 ///     println!("{}", String::from_utf8(buffer.clone()).unwrap());
 ///     Ok(())
-/// }
+/// # }
 /// ```
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, Default)]
 pub struct PrometheusLayer {
-    registry: Registry,
-    operation_duration_seconds_buckets: Vec<f64>,
-    operation_bytes_buckets: Vec<f64>,
-    path_label_level: usize,
+    interceptor: PrometheusInterceptor,
 }
 
 impl PrometheusLayer {
     /// Create a [`PrometheusLayer`] while registering itself to this registry.
-    pub fn new(registry: Registry) -> Self {
-        Self {
-            registry,
-            operation_duration_seconds_buckets: exponential_buckets(0.01, 2.0, 16).unwrap(),
-            operation_bytes_buckets: exponential_buckets(1.0, 2.0, 16).unwrap(),
-            path_label_level: 0,
-        }
+    pub fn new(registry: &Registry) -> Self {
+        let interceptor = PrometheusInterceptor::default();
+        Self { interceptor }.register(registry)
+    }
+
+    /// Register the metrics into the registry.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use log::debug;
+    /// # use opendal::layers::PrometheusLayer;
+    /// # use opendal::services;
+    /// # use opendal::Operator;
+    /// # use opendal::Result;
+    /// #
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<()> {
+    ///     // Pick a builder and configure it.
+    ///     let builder = services::Memory::default();
+    ///     let registry = prometheus::default_registry();
+    ///
+    ///     let op = Operator::new(builder)
+    ///         .expect("must init")
+    ///         .layer(PrometheusLayer::default().register(registry))
+    ///         .finish();
+    ///     debug!("operator: {op:?}");
+    ///
+    ///     Ok(())
+    /// # }
+    /// ```
+    pub fn register(self, registry: &Registry) -> Self {
```

Comment:
We don't need this API. `PrometheusLayer::new()` should be enough.

Reply:
IMO, `PrometheusLayer::new` should just be a helper function that uses the default config and registers the metrics with the default config to the registry.

Reply:
Personally, I don't believe we should perform `register` during `new` unless the upstream API forbid it. Creating the layer itself should have no side effects. Everything should occur during the layer.

So `PrometheusLayer::new()` should just take a `registry` and return `Self`. We will perform the registry during `Layer::layer`.

Reply:
So how about `PrometheusClientLayer` in #5073

Reply:
Okay, how about this: Let's use your previous proposed builder's API.

```rust
let layer: impl Layer = PrometheusClientLayer::builder()
    .operation_bytes_buckets(vec![1.0, 2.0])
    .register(registry);
```

`PrometheusClientLayer::builder()` returns a `PrometheusClientLayerBuilder`. Users can only use this layer after calling `register`. We can provide `default()` if the upstream lib has `default` registry concept.

What do you think?

Reply:
So do we need to use similar APIs in `PrometheusLayer`?

Reply:
> So do we need to use similar APIs in `PrometheusLayer`?

I believe this design can adapt to all metrics libraries. What do you think?

Reply:
Yes, if you agree with the design, I'll apply it to `PrometheusLayer` and `PrometheusClientLayer`.
But I'm not sure if this design will fit well with other metric layers, I'm not very optimistic after using the `metrics` library.

Reply:
I feel like we can call `with_recorder` direcly: https://docs.rs/metrics/latest/src/metrics/recorder/mod.rs.html#126